src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts

```diff
@@ -0,0 +1,36 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as extensionsRegistry from 'vs/workbench/services/extensions/common/extensionsRegistry';
+import { ITerminalTypeContribution, ITerminalContributions, terminalContributionsDescriptor } from 'vs/workbench/contrib/terminal/common/terminal';
+import { flatten } from 'vs/base/common/arrays';
+import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
+
+// terminal extension point
+export const terminalsExtPoint = extensionsRegistry.ExtensionsRegistry.registerExtensionPoint<ITerminalContributions>(terminalContributionsDescriptor);
+
+export interface ITerminalContributionService {
+	readonly _serviceBrand: undefined;
+
+	readonly terminalTypes: ReadonlyArray<ITerminalTypeContribution>;
+}
+
+export const ITerminalContributionService = createDecorator<ITerminalContributionService>('terminalContributionsService');
+
+export class TerminalContributionService implements ITerminalContributionService {
+	public readonly _serviceBrand = undefined;
+
+	private _terminalTypes: ReadonlyArray<ITerminalTypeContribution> = [];
+
+	public get terminalTypes() {
+		return this._terminalTypes;
+	}
+
+	constructor() {
+		terminalsExtPoint.setHandler(contributions => {
```

Comment:
We need to protect usage of this for only when proposed API is enabled

Reply:
Seems like the common pattern in vscode itself is to silently ignore contributions that require a proposed API to work, e.g. `themeExtensionPoints.ts` or `proxyResolver.ts`. Not sure if that's the best thing for extension authors, but followed that pattern here.